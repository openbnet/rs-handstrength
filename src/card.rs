#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub enum Suit {
    S,
    H,
    C,
    D,
    A, // any *
    E, // TrainerSuit first suit in cards
    F, //2nd suit
    G, // 3rd suit
    X, // Except E - 3 possible
    Y, // Except E and F - 2 possible
    Z, // Except E, F, and G -4th suit

}

#[derive(Clone, Copy, Debug, Eq, Hash,PartialOrd, PartialEq, Ord)]
pub struct Card {
    pub value: u8,
    pub suit: Suit,
}
use std::fmt;

impl fmt::Display for Suit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let suit = match self {
            Suit::C => "c",
            Suit::D => "d",
            Suit::H => "h",
            Suit::S => "s",
            Suit::A => "*",
            Suit::E => "e",
            Suit::F => "f",
            Suit::G => "g",
            Suit::X => "x",
            Suit::Y => "y",
            Suit::Z => "z",
        };
        write!(f, "{}", suit)
    }
}

impl fmt::Display for Card {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} of {}", self.value, self.suit)
    }
}
impl Default for Card {
    fn default() -> Self {
        Card {
            value: 0,
            suit: Suit::S,
        }
    }
}