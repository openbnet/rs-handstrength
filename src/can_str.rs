use crate::card::{Card, Suit};
use crate::sorting::get_same_value_map;
use crate::can_libs::*;
use itertools::Itertools;
use std::collections::HashSet;

pub fn can_have_str(comcards: &Vec<Card>) -> CanHaveCombis {
    let mut can_have_combis: CanHaveCombis = Vec::new();
    let grouped_cards = get_same_value_map(comcards);
    let unique_cards: Vec<Card> = grouped_cards.iter()
        .filter(|group| group.len() == 1)
        .map(|group| group[0].clone())
        .collect();
    let card_set: HashSet<u8> = unique_cards.iter().map(|card| card.value).collect();

    for start in [10, 1, 9, 8, 7, 6, 5, 4, 3, 2] {
        let straight_values = (start..start + 5).collect::<Vec<u8>>();

        let board_candidates: Vec<Card> = unique_cards.iter()
            .filter(|&card| straight_values.contains(&card.value))
            .cloned()
            .collect();

        let missing_values: Vec<u8> = straight_values.iter()
            .filter(|&val| !card_set.contains(val))
            .cloned()
            .collect();

        let possible_hands_combinations = match board_candidates.len() {
            5 => {
                // All five cards form a straight
                vec![board_candidates.iter().combinations(2).map(|combo| {
                    combo.into_iter().cloned().collect::<Vec<Card>>()
                }).collect::<Vec<Vec<Card>>>()] // This is a Vec<SameRankHands>
            },
            4 => {
                // Pair the missing value with each card in the straight
                let missing_card = Card { value: missing_values[0], suit: Suit::A };
                vec![board_candidates.iter()
                    .filter(|card| card.value != missing_values[0])
                    .map(|&existing_card| vec![missing_card.clone(), existing_card])
                    .collect::<Vec<Vec<Card>>>()] // This is a Vec<SameRankHands>
            },
            3 => {
                // Create combinations with the two missing values
                vec![missing_values.iter().combinations(2).map(|pair| {
                    pair.into_iter().map(|&val| Card { value: val, suit: Suit::A }).collect::<Vec<Card>>()
                }).collect::<Vec<Vec<Card>>>()] // This is a Vec<SameRankHands>
            },
            _ => Vec::new()
        };
        
        if !possible_hands_combinations.is_empty() {
            can_have_combis.push((board_candidates, possible_hands_combinations));
        }
    }

    can_have_combis
}
