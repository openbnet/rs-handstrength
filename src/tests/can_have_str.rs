#[cfg(test)]
mod tests {
    use crate::card::{Card, Suit};
    use crate::can_str::can_have_str;
    use insta::assert_debug_snapshot;
    #[test]
    fn cannot_str() {
        let cards = vec![
            Card { value: 2, suit: Suit::S },
            Card { value: 10, suit: Suit::S },
            Card { value: 11, suit: Suit::H },
            Card { value: 10, suit: Suit::D },
            Card { value: 5, suit: Suit::C },
        ];

        let str_combis = can_have_str(&cards);
        // println!("fs_combinations, {:?}", str_combis);
        // Check for 13 possible combinations (excluding 10)
        assert_eq!(str_combis.len(), 0);
    }
    #[test]
    fn can_str1() {
        let cards = vec![
            Card { value: 1, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::H },
            Card { value: 8, suit: Suit::C },
            Card { value: 9, suit: Suit::C },
        ];

        let str_combis = can_have_str(&cards);
        // println!("fs_combinations, {:?}", str_combis);
        // Check for 13 possible combinations (excluding 10)
        assert_eq!(str_combis.len(), 1);
        assert_debug_snapshot!(str_combis);
    }

    #[test]
    fn can_str2() {
        let cards = vec![
            Card { value: 1, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::H },
            Card { value: 4, suit: Suit::C },
            Card { value: 9, suit: Suit::C },
        ];

        let str_combis = can_have_str(&cards);
        // println!("fs_combinations, {:?}", str_combis);
        // println!("test len {:?}", cards.len());
        // Check for 13 possible combinations (excluding 10)
        assert_eq!(str_combis.len(), 2);
        assert_debug_snapshot!(str_combis);
    }
    #[test]
    fn can_str3() {
        let cards = vec![
            Card { value: 1, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::H },
            Card { value: 4, suit: Suit::C },
            Card { value: 5, suit: Suit::C },
        ];

        let str_combis = can_have_str(&cards);
        // println!("fs_combinations, {:?}", str_combis);
        // println!("test len {:?}", cards.len());
        // Check for 13 possible combinations (excluding 10)
        assert_eq!(str_combis.len(), 3);
        assert_debug_snapshot!(str_combis);
    }
    #[test]
    fn can_str4() {
        let cards = vec![
            Card { value: 1, suit: Suit::S },
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::H },
            Card { value: 10, suit: Suit::C },
            Card { value: 11, suit: Suit::C },
        ];

        let str_combis = can_have_str(&cards);
        // println!("fs_combinations, {:?}", str_combis);
        assert_eq!(str_combis.len(), 2);
        assert_debug_snapshot!(str_combis);
    }
    #[test]
    fn can_str5() {
        let cards = vec![
            Card { value: 10, suit: Suit::S },
            Card { value: 9, suit: Suit::S },
            Card { value: 7, suit: Suit::H },
            Card { value: 6, suit: Suit::C },
            Card { value: 5, suit: Suit::C },
        ];

        let str_combis = can_have_str(&cards);
        // println!("fs_combinations, {:?}", str_combis);
        assert_eq!(str_combis.len(), 5);
        assert_debug_snapshot!(str_combis);
    }
}