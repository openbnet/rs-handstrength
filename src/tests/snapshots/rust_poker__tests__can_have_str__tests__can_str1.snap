---
source: src/tests/can_have_str.rs
expression: str_combis
---
[
    (
        [
            Card {
                value: 3,
                suit: H,
            },
            Card {
                value: 2,
                suit: S,
            },
            Card {
                value: 1,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 4,
                        suit: A,
                    },
                    Card {
                        value: 5,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
]
